Merge Labs 3.X into dev

quinchs opened the pull request:
# Summary
This PR merges our work at [Discord.Net Labs](https://github.com/Discord-Net-Labs/Discord.Net-Labs) over to DNET.
There are a lot of changes to go thru so I've done my best to link them below. **Note** the links to the files will take you to our `Refactor/3.x` branch on labs, that branch is up to date with this PR and is what future comits will go to.

This PR is going to be marked as draft as there is a lot of stuff to go thru and I can presume some stuff will need to be changes in order to meet some standards within the library.

## Major changes

-   Added Interaction Support.
-   Added Application commands (slash, user, message).
-   Added Message Components (buttons, select menus).
-   Added Thread Channels.
-   Added Stage Channels.
-   Revamped Stickers.

## Minor changes

-   Added `TimestampTag`.
-   Made `Hierarchy` a `IGuildUser` property.
-   Changes embed discription length to 4096.
-   Added `Name` property to teams.
-   Added url validation to embeds.
-   [Added `NsfwLevel` to Guilds.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Guilds/NsfwLevel.cs)
-   Added helpers to `Emoji` for parsing.
-   Fixed gateway serialization to include nulls.
-   Added banner and accent color to guild users.
-   Fixed `CurrentUserId` in sharded clients being null.
-   Fixed Guild owner and Admin `GuildPermissions.All`.
-   Added `RatelimitCallback` to `RequestOptions`.

### Interactions

Interactions are the backbone of most of the new features in this PR. Here's a quick overview of the supported features:

-   [x] Recieve interaction events from the gateway
-   [x] Recieve interaction events via webhook client
-   [x] Respond to interaction events using responses and followup messages
-   [x] Ephemeral message.
-   [x] Responses with multiple embeds.
-   [x] followup with files.

<details>
<summary>Discord.Net.Core</summary>

-   [Added `IDiscordInteraction` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/IDiscordInteraction.cs)
-   [Added `IDiscordInteractionData` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/IDiscordInteractionData.cs)
-   [Added `InteractionResponseType` enum.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/InteractionResponseType.cs)
-   [Added `InteractionType` enum.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/InteractionType.cs)
</details>

<details>
<summary>Discord.Net.WebSocket</summary>

-   [Added `SocketInteraction` base class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/SocketInteraction.cs)
-   Added `InteractionCreated` event.
</details>

<details>
<summary>Discord.Net.Rest</summary>

-   [Added `InteractionHelper` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Entities/Interactions/InteractionHelper.cs)
-   [Added `Interaction` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/IDiscordInteraction.cs)
-   [Added `InteractionConverter` JsonConverter.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Net/Converters/InteractionConverter.cs)
-   [Added `InteractionCallbackData` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/InteractionCallbackData.cs)
-   [Added `InteractionResponse` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/InteractionResponse.cs)
</details>

### Application Commands

Application commands are the staple of this PR. Here's a list of supported features:

-   [x] Create guild/global commands.
-   [x] Recieve command interactions via gateway.
-   [x] Modify commands.
-   [x] Change command permissions.
-   [x] Context commands.
-   [x] Slash commands.
-   [x] Bulk overwrite guild/global commands
-   [x] Bulk overwrite command permissions.
-   [x] Application command builders.
-   [x] Support for receiving & responding to application command based interactions.
-   [x] Autocomplete application commands.

<details>
<summary>Discord.Net.Core</summary>

-   [Added `IApplicationCommand` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/IApplicationCommand.cs)
-   [Added `IApplicationCommandInteractionData` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/IApplicationCommandInteractionData.cs)
-   [Added `IApplicationCommandInteractionDataOption` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/IApplicationCommandInteractionDataOption.cs)
-   [Added `IApplicationCommandOption` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOption.cs)
-   [Added `IApplicationCommandOptionChoice` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/IApplicationCommandOptionChoice.cs)
-   [Added `ApplicationCommandProperties` base class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/ApplicationCommandProperties.cs)
-   Added `ApplicationCommandType` enum.
-   [Added `ApplicationCommandOptionType` enum.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/ApplicationCommandOptionType.cs)
-   Added `ApplicationCommandOptionChoiceProperties` class.
-   Added `ApplicationCommandOptionProperties` class.
-   [Added `MessageCommandBuilder` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/Context%20Menus/MessageCommandBuilder.cs)
-   [Added `MessageCommandProperties` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/Context%20Menus/MessageCommandProperties.cs)
-   [Added `UserCommandBuilder` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/Context%20Menus/UserCommandBuilder.cs)
-   [Added `UserCommandProperties` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/Context%20Menus/UserCommandProperties.cs)
-   [Added `SlashCommandBuilder` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/SlashCommandBuilder.cs)
-   [Added `SlashCommandProperties` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/SlashCommandProperties.cs)
-   Added `ApplicationCommandPermission` class.
-   [Added `ApplicationCommandPermissionTarget` enum.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Permissions/ApplicationCommandPermissionTarget.cs)
-   Added `GuildApplicationCommandPermission` class.
</details>

<details>
<summary>Discord.Net.WebSocket</summary>

-   [Added `SocketApplicationCommand` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketApplicationCommand.cs)
-   [Added `SocketApplicationCommandChoice` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketApplicationCommandChoice.cs)
-   [Added `SocketApplicationCommandOption` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketApplicationCommandOption.cs)
-   [Added `SocketCommandBase` base class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketCommandBase.cs)
-   [Added `SocketCommandBaseData<TOption>` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketCommandBaseData.cs)
-   [Added `SocketResolvableData<T>` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketResolvableData.cs)
-   [Added `SocketMessageCommand` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/Context%20Menu%20Commands/Message%20Commands/SocketMessageCommand.cs)
-   [Added `SocketMessageCommandData` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/Context%20Menu%20Commands/Message%20Commands/SocketMessageCommandData.cs)
-   [Added `SocketUserCommand` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/Context%20Menu%20Commands/User%20Commands/SocketUserCommand.cs)
-   [Added `SocketUserCommandData` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/Context%20Menu%20Commands/User%20Commands/SocketUserCommandData.cs)
-   [Added `SocketSlashCommand` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/Slash%20Commands/SocketSlashCommand.cs)
-   [Added `SocketSlashCommandData` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/Slash%20Commands/SocketSlashCommandData.cs)
-   [Added `SocketSlashCommandDataOption` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/Slash%20Commands/SocketSlashCommandDataOption.cs)
-   Added `SlashCommandExecuted` event to socket client.
-   Added `UserCommandExecuted` event to socket client.
-   Added `MessageCommandExecuted` event to socket client.
-   [Added `SocketApplicationCommand` to the state class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketApplicationCommand.cs)
-   Added methods to get/create/modify/delete guild commands to socket guild.
-   Added methods to get/create/modify/delete global commands to SocketClient.
</details>

<details>
<summary>Discord.Net.Rest</summary>

-   [Added `RestApplicationCommand` base class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Entities/Interactions/RestApplicationCommand.cs)
-   [Added `RestApplicationCommandChoice` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Entities/Interactions/RestApplicationCommandChoice.cs)
-   [Added `RestApplicationCommandOption` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Entities/Interactions/RestApplicationCommandOption.cs)
-   [Added `RestGlobalCommand` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Entities/Interactions/RestGlobalCommand.cs)
-   [Added `RestGuildCommand` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Entities/Interactions/RestGuildCommand.cs)
-   Added methods to get/create/modify/delete commands into the RestClient.
-   Added methods to get/create/modify/delete guild commands to rest guild.
-   [Added `ApplicationCommand` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/IApplicationCommand.cs)
-   [Added `ApplicationCommandInteractionData` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/IApplicationCommandInteractionData.cs)
-   [Added `ApplicationCommandInteractionDataOption` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/IApplicationCommandInteractionDataOption.cs)
-   [Added `ApplicationCommandInteractionDataResolved` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/ApplicationCommandInteractionDataResolved.cs)
-   [Added `ApplicationCommandOption` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/ApplicationCommandOption.cs)
-   [Added `ApplicationCommandOptionChoice` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/ApplicationCommandOptionChoice.cs)
-   [Added `ApplicationCommandPermissions` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Permissions/ApplicationCommandPermissions.cs)
-   Added `GuildApplicationCommandPermission` model.

</details>

### Message Components

Message components are another big part in the interaction additions,Here's a quick overview of the supported features:

-   [x] Buttons
-   [x] Select Menu's
-   [x] Builders
-   [x] Support for sending with messages
-   [x] Support for recieving & responding to component based interactions

<details>
<summary>Discord.Net.Core</summary>

-   [Added `IMessageComponent` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/Message%20Components/IMessageComponent.cs)
-   [Added `MessageComponent` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/Message%20Components/IMessageComponent.cs)
-   [Added `ButtonComponent` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/ButtonComponent.cs)
-   [Added `ButtonStyle` enum.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/Message%20Components/ButtonStyle.cs)
-   [Added `SelectMenuComponent` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/SelectMenuComponent.cs)
-   [Added `SelectMenuOption` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/SelectMenuOption.cs)
-   [Added `ActionRowComponent` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/ActionRowComponent.cs)
-   [Added `ComponentBuilder` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/Message%20Components/ComponentBuilder.cs)
-   [Added `ComponentType` enum.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/Message%20Components/ComponentType.cs)
-   [Added `MessageComponent` field to all send message functions.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/Message%20Components/IMessageComponent.cs)
-   Added `IReadOnlyCollection<IMessageComponent>` property to `IMessage` and its inheriting classes.
-   [Added `MessageComponent` to `MessageProperties` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Interactions/Message%20Components/IMessageComponent.cs)
</details>

<details>
<summary>Discord.Net.WebSocket</summary>

-   [Added `SocketMessageComponent` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/Message%20Components/SocketMessageComponent.cs)
-   [Added `SocketMessageComponentData` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Interaction/Message%20Components/SocketMessageComponentData.cs)
-   Added `SelectMenuExecuted` event to socket client.
-   Added `ButtonExecuted` event to socket client.
</details>

<details>
<summary>Discord.Net.Rest</summary>

-   Added functions to `MessageHelper` class to help with components.
-   [Added `MessageComponentConverter` JsonConverter.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Net/Converters/MessageComponentConverter.cs)
-   [Added `ButtonComponent` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/ButtonComponent.cs)
-   [Added `ActionRowComponent` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/ActionRowComponent.cs)
-   [Added `SelectMenuComponent` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/SelectMenuComponent.cs)
-   [Added `SelectMenuOption` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/SelectMenuOption.cs)
-   [Added `MessageComponentInteractionData` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/MessageComponentInteractionData.cs)
</details>

### Thread Channels

Thread channels are a new form of channels that act similar to text channels. Heres the list of features:

-   [x] Create thread channels.
-   [x] Recieve thread channel interactions/messages via gateway.
-   [x] Modify thread channels.
-   [x] Archive/unarchive thread channels.

<details>
<summary>Discord.Net.Core</summary>

-   [Added `IThreadChannel` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Channels/IThreadChannel.cs)
-   [Added `ThreadType` enum.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Channels/ThreadType.cs)
-   [Added `ThreadArchiveDuration` enum.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Channels/ThreadArchiveDuration.cs)
-   Added properties into `TextChannelProperties` for modifying threads.
-   Added `CreateThreadAsync` to `ITextChannel` interface.
-   Added `HasThread` to `MessageFlags` enum.
-   Added `ManageThreads`, `UsePublicThreads`, and `UsePrivateThreads` to `GuildPermissions`.
-   Added `GetThreadChannelAsync` and `GetThreadChannelsAsync` to `IGuild`.
</details>

<details>
<summary>Discord.Net.WebSocket</summary>

-   [Added `SocketThreadChannel` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Channels/SocketThreadChannel.cs)
-   [Added `SocketThreadUser` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Users/SocketThreadUser.cs)
-   Added `ThreadChannels` property to `SocketGuild` class.
-   Added `CreateThreadAsync` to `SocketTextChannel`.
-   [Added `Threads` property to `ExtendedGuild` gateway model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/ChannelThreads.cs)
-   Added `ThreadListSync` gateway model.
-   Added `ThreadCreated`, `ThreadUpdated`, `ThreadDeleted`, `ThreadMemberJoined`, and `ThreadMemberLeft` events to socket client.
-   Added case for thread channels in `SocketGuildChannel`.
-   [Added `Threads` property in `SocketTextChannel`.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/ChannelThreads.cs)
-   [Added `ThreadMembersUpdate` gateway model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/API/Gateway/ThreadMembersUpdate.cs)

</details>

<details>
<summary>Discord.Net.Rest</summary>

-   Added thread related properties to `Channel` model.
-   [Added `ChannelThreads` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/ChannelThreads.cs)
-   [Added `Threads` property to `Guild` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/ChannelThreads.cs)
-   [Added `ThreadMember` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/ThreadMember.cs)
-   [Added `ThreadMetadata` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/ThreadMetadata.cs)
-   [Added `StartThreadParams` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Rest/StartThreadParams.cs)
-   Added relative methods to RestClient.
-   [Added `ModifyThreadParams` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Rest/ModifyThreadParams.cs)
-   Added `CreateThreadAsync` to `RestTextChannel`.
-   [Added `RestThreadChannel` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Entities/Channels/RestThreadChannel.cs)
-   [Added `ThreadHelper` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Entities/Channels/ThreadHelper.cs)
-   [Added `RestThreadUser` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Entities/Users/RestThreadUser.cs)
-   Added `GetThreadChannelAsync` to `RestGuild`.
-   Added `GetThreadChannelsAsync` to `RestGuild`.
-   Added case for thread channels in `SocketGuildChannel`.
</details>

### Stage channels

Stage channels are a new form of channels that act similar to voice channels. Heres the list of features:

-   [x] Creating stage channels.
-   [x] Starting/stopping/modifying stage instances
-   [x] Adding/removing speakers
-   [x] Becoming speaker

<details>
<summary>Discord.Net.Core</summary>

-   [Added `IStageChannel` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Channels/IStageChannel.cs)
-   [Added `StagePrivacyLevel` enum.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Channels/StagePrivacyLevel.cs)
-   [Added `StageInstanceProperties` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Channels/StageInstanceProperties.cs)
-   Added `RequestToSpeakTimestamp` to `IVoiceState`.
-   Added `GetStageChannelAsync` and `GetStageChannelsAsync` to `IGuild`.
</details>

<details>
<summary>Discord.Net.WebSocket</summary>

-   [Added `SocketStageChannel` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Channels/SocketStageChannel.cs)
-   [Added `StageInstance` gateway model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/StageInstance.cs)
-   Added `StageStarted`, `StageEnded`, `StageUpdated`, `RequestToSpeak`, `SpeakerAdded`, and `SpeakerRemoved` events to socket client.
-   Added `GetStageChannel` to `SocketGuild`.
-   Added case for stage channels in `SocketGuildChannel`.
-   Added `RequestToSpeakTimestamp` to `SocketVoiceState`.
-   Added `StageChannels` property to `SocketGuild`.

</details>

<details>
<summary>Discord.Net.Rest</summary>

-   [Added `RestStageChannel` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/Entities/Channels/RestStageChannel.cs)
-   Added `RequestToSpeakTimestamp` to `VoiceState` model.
-   [Added `CreateStageInstanceParams` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Rest/CreateStageInstanceParams.cs)
-   [Added `ModifyStageInstanceParams` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Rest/ModifyStageInstanceParams.cs)
-   Added relative methods to RestClient.
-   Added `ModifyAsync` for stage channels to `ChannelHelper.
-   Added case for stage channels in `RestGuildChannel`.
-   Added `GetStageChannelAsync` and `GetStageChannelsAsync` to `RestGuild`.
-   [Added `ModifyVoiceStateParams` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Rest/ModifyVoiceStateParams.cs)
-   [Added `StageInstance` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/StageInstance.cs)
</details>

### Stickers

Stickers we're initially supported in dnet but discord has changed how they work. Heres the list of features:

-   [x] Sending stickers with messages
-   [x] Creating stickers
-   [x] Default stickers & Sticker packs
-   [x] Modifying and deleting stickers

<details>
<summary>Discord.Net.Core</summary>

-   [Added `ISticker` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Stickers/ISticker.cs)
-   [Added `ICustomSticker` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Stickers/ICustomSticker.cs)
-   [Added `IStickerItem` interface.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Stickers/IStickerItem.cs)
-   [Added `StickerPack` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Stickers/StickerPack.cs)
-   [Added `StickerProperties` class.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Stickers/StickerProperties.cs)
-   [Added `StickerType` enum.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Stickers/StickerType.cs)
-   Added `ISticker[]` parameter to `SendMessageAsync` functions.
-   Added `GetStickerUrl` to `CDN`.
-   Added `Stickers` property to `IGuild`.
-   Added `CreateStickerAsync`, `GetStickerAsync`, `GetStickersAsync`, and `DeleteStickerAsync` to `IGuild`.
-   Added

</details>

<details>
<summary>Discord.Net.WebSocket</summary>

-   [Added `SocketSticker` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Stickers/SocketSticker.cs)
-   [Added `SocketCustomSticker` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Stickers/SocketCustomSticker.cs)
-   [Added `SocketUnknownSticker` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/Entities/Stickers/SocketUnknownSticker.cs)
-   Added `Stickers` property to `SocketGuild`.
-   Added `GetStickerAsync`, `GetStickersAsync`, and `CreateStickerAsync` to `SocketGuild`.
-   Added `DefaultStickerPacks` property to `DiscordSocketClient` and `DiscordShardedClient`. When the client starts it will automatically download the default sticker packs, if running shared then the sharded client will download them and pass the packs to the individual clients.
-   Added `GetStickerAsync` to `BaseSocketClient`.
-   Added `Stickers` property to `SocketMessage`.
-   [Added `GuildStickerUpdateEvent` model
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.WebSocket/API/Gateway/GuildStickerUpdateEvent.cs)
-   Added `GuildStickerCreated`, `GuildStickerUpdated`, and `GuildStickerDeleted` events to socket client.
</details>

<details>
<summary>Discord.Net.Rest</summary>

-   [Added `StickerItem` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Stickers/IStickerItem.cs)
-   [Added `StickerItem` to `Message` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Stickers/IStickerItem.cs)
-   [Added `NitroStickerPacks` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Common/NitroStickerPacks.cs)
-   [Added `User` model to `Sticker` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Users/IGroupUser.cs)
-   [Added `StickerPack` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Stickers/StickerPack.cs)
-   [Added `CreateStickerParams` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Rest/CreateStickerParams.cs)
-   [Added `ModifyStickerParams` model.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Rest/API/Rest/ModifyStickerParams.cs)
-   Added relative methods to RestClient.
-   Added `CreateStickerAsync`, `ModifyStickerAsync` and `DeleteStickerAsync` to `GuildHelper`
-   [Added `StickerItem` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Stickers/IStickerItem.cs)
-   [Added `CustomSticker` entity.
](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/refactor/3.x/src/Discord.Net.Core/Entities/Stickers/ICustomSticker.cs)
-   Added `GuildId` property to `Sticker` model.
-   Added `Stickers` property to `RestGuild`.
-   Added `CreateStickerAsync`, `GetStickerAsync`, `GetStickersAsync`, and `DeleteStickerAsync` to `RestGuild`.
-   Added `Stickers` property to `RestMessage`
</details>

quinchs commented:
Related issues:
#1915 #1909 #1899 #1897 #1856 #1821 #1809 #1755 #1716 #1733 #1717 #1922

quinchs committed:
meta: bump version

killerfrienddk committed:
Null or empty fix (#176)

* Add components and stickers to ReplyAsync extension

* Fixed null or empty

* Changed Label to Description

* -||-

Co-authored-by: quin lynch <lynchquin@gmail.com>

killerfrienddk committed:
More regions (#177)

* Preconditions

* ChannelHelper

* RestDMChannel

* RestGroupChannel

* RestBan

* RestGroupUser

* EntityExtensions

* DiscordSocketClient

* DiscordSocketClient

killerfrienddk committed:
Discord.net.core.xml fix (#178)

* Changed Label to Description

* Added Discord- .MessageComponent .ISticker[]

,Discord.MessageComponent,Discord.ISticker[] to ReplyAsync

quinchs committed:
Remove references to labs

quinchs committed:
Update Discord.Net.sln

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into merger-labs

drobbins329 committed:
Added SendMessagesInThreads and StartEmbeddedActivities. (#175)

* Added SendMessagesInThreads and StartEmbeddedActivities.

Adjusted owner perms.
Change UsePublicThreads -> CreatePublicThreads
Change UsePrivateThreads -> CreatePrivateThreads

* removed extra ///

* Added UsePublicThreads and UsePrivateThreads back with Obsolete Attribute

* removed 'false' from Obsolete Attribute

quinchs committed:
Squashed commit of the following:

commit dca41a348e36a9b4e7006ef3a76377eb32aad276
Author: quin lynch <lynchquin@gmail.com>
Date:   Thu Sep 23 07:02:19 2021 -0300

    Autocomplete commands

quinchs committed:
meta: xml. closes #171

quinchs committed:
Revert user agent and $device to dnet

quinchs committed:
Merge branch 'refactor/3.x' into release/3.x

quinchs committed:
meta: bump version

quinchs committed:
meta: bump vers

quinchs committed:
Fix sticker args

killerfrienddk committed:
Grammer fix (#179)

quinchs committed:
Made IVoiceChannel mentionable

killerfrienddk committed:
Embeds array for send message async (#181)

* meta: bump version

* meta: bump vers

* Fix sticker args

* Grammer fix (#179)

* Added embeds for SendMessageAsync

* [JsonProperty("embed")] forgot to remove this

 public Optional<Embed> Embed { get; set; }

* It has been done as requested.

* Changed the old way of handeling single embeds

* Moved embeds param and added options param

* xmls

Co-authored-by: quin lynch <lynchquin@gmail.com>

MrCakeSlayer committed:
Fix thread permissions (#183)

* Update GuildPermissionsTests.cs

* Update GuildPermissions.cs

killerfrienddk committed:
Use compound assignment (#186)

* Used compound assignment

* -||-

* -||-

killerfrienddk committed:
Remove unnecessary suppression (#188)

killerfrienddk committed:
Inlined variable declarations (#185)

killerfrienddk committed:
Fixed some warnings (#184)

* Fixed some warnings

* Another fixed warning

* Changed the SSendFileAsync to SendFileAsync

* Removed para AlwaysAcknowledgeInteractions

* Moved it back to the previous version

* Added periods to the end like quin requested!! :((

Co-authored-by: MrCakeSlayer <13650699+MrCakeSlayer@users.noreply.github.com>

killerfrienddk committed:
Object initialization can be simplified fixed (#189)

killerfrienddk committed:
Conditional-expression-simplification (#193)

killerfrienddk committed:
Capitlazation fixes (#192)

killerfrienddk committed:
Removed-this. (#191)

killerfrienddk committed:
Use 'switch' expression (#187)

* Use 'switch' expression

* Reverted it to the old switch case

killerfrienddk committed:
Fixed-compiler-error (#194)

drobbins329 committed:
Submitting updates to include new permissions. (#195)

* Submitting updates to include new permissions.

* Make old permissions obsolete and update tests

Co-authored-by: quin lynch <lynchquin@gmail.com>

quinchs committed:
Update azure-pipelines.yml

quinchs committed:
Update azure-pipelines.yml

quinchs committed:
Update azure-pipelines.yml

quinchs committed:
Add support for long in autocomplete option

d4n3436 committed:
Add support for sending files with multiple embeds (#196)

* Add support for sending files with multiple embeds

* Simplify prepending single embed to embed array

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into merger-labs

quinchs commented:
[This comit](https://github.com/discord-net/Discord.Net/pull/1923/commits/17b5da2d577dc3db6a7f156709d6ead826daa142) includes a lot of refactor to other parts of the library to simply code as well as all the grammar fixes

killerfrienddk committed:
Consistency for embeds endpoints (#197)

* Changed the way of handling prepending of embeds.

For consistency.

* reformatted the summary

quinchs committed:
Revert pipeline

quinchs committed:
Fix duplicate merge conflicts

WilliamWelsh committed:
Changed minimum slash command name length to 1 per Discord API docs (#198)

killerfrienddk committed:
Channel endpoints requirements correction (#199)

* Added some requirements to channels for topic

* Changed check from NotNullOrEmpty to NotNullOrEmpty

* Added some requirements to channels for name

Preconditions.LessThan

* Formatting of file

killerfrienddk committed:
Added restriction for description not being null (#200)

quinchs committed:
Update azure-pipelines.yml

quinchs committed:
Update deploy.yml

quinchs committed:
Remove version tag from proj

quinchs committed:
Update deploy.yml

quinchs committed:
Removed versions from project files

killerfrienddk committed:
Removed style of the nuget badge and added logo (#201)

The style was not properly added to it and the plastic version does not look good with the discord badge.
I thought it would look better with a logo

quinchs committed:
Fix Type not being set in SocketApplicationCommand

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
Remove useless GuildId property

quinchs committed:
meta: update XML

quinchs committed:
Add Autocomplete to SlashCommandOptionBuilder

Commit:
Added autocomplete in SlashCommandOptionBuilder. (#206)

Co-authored-by: Quin Lynch <49576606+quinchs@users.noreply.github.com>

quinchs committed:
Fix duplicate autocomplete

quinchs committed:
Fix #208

quinchs committed:
Fix sub commands being interpreted as a parameter for autocomplete

quinchs committed:
Fix exposed optional

CottageDwellingCat committed:
Support the discord:// protocol in buttons (#207)

* Update UrlValidation.cs

* Update ComponentBuilder.cs

* Add docs and better error messages.

* Fix wonky intentation

MrCakeSlayer committed:
Add competing activity status type (#205)

* Update GuildPermissionsTests.cs

* Update GuildPermissions.cs

* Add competing status type

WilliamWelsh committed:
Add Icons to IRole (#204)

* Added icon field to IRole

* Added GetGuildRoleIconUrl()

emillly-b committed:
Added Clean Content Function (#174)

* Added Clean Content Function

* Fixed Spelling problems and bad var handling

* Add StripMarkDown Method

emillly-b committed:
Clean Content Expanded (#212)

* Implement CleanContent In IMessage & RestMessage

* Update Spelling and Documentation

* Add SanatizeMessage to MessageHelper and Refactor Rest and Socket Message

marens101 committed:
Add event for autocomplete interaction (#214)

killerfrienddk committed:
Spelling corrections (#215)

quinchs committed:
Remove null collections

killerfrienddk committed:
Followup with file async warnings (#216)

* Changed from NotNullOrWhitespace to NotNullOrEmpty

* Added NotNullOrEmpty on filename

* Added system to interpret from the path

* Added a check for if it contains a period

* It has been done, how ever it will break stuff

* Changed to use ??= how ever still added error check

* Added space under check

* Changed from with a period to valid file extension

* Added checks for SendFileAsync

* Removed filename != null &&

CottageDwellingCat committed:
Add channel types in application command options. (#217)

* add channel types in application command options

* Indent Docs

quinchs committed:
Stage instance audit logs as well as thread audit log type

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
Update azure-pipelines.yml

quinchs committed:
Update azure-pipelines.yml

quinchs committed:
Fix system messages not including mentioned users. Added ContextMenuCommand message type

d4n3436 committed:
Remove file extension check (#218)

quinchs committed:
Fix NRE in modify guild channel

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
Fix 429's not being accounted for in ratelimit updates

quinchs committed:
meta: add net5 framework

Co-Authored-By: MrCakeSlayer <13650699+MrCakeSlayer@users.noreply.github.com>

MrCakeSlayer committed:
Proper doc logos (#221)

* Update GuildPermissionsTests.cs

* Update GuildPermissions.cs

* Add competing activity status type

* logo changes

* logo text as path

* add missing logo

quinchs committed:
Update package logo and favicon

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
Update docfx references

quinchs committed:
Remove XML files and use original pipeline format

quinchs committed:
Remove console writeline

quinchs committed:
Remove Console.WriteLine

quinchs committed:
Remove useless log

quinchs committed:
Rename Available sticker field to IsAvailable

quinchs committed:
Rename Available to IsAvailable in stickers

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
Add summary indent for role members

quinchs committed:
Add summary indent to SocketInvite

quinchs committed:
Rename DefaultPermission to IsDefaultPermission

quinchs committed:
Rename Default to IsDefault and Required to IsRequired in IApplicationCommandOption

quinchs committed:
Rename Default and Required to IsDefault and IsRequired in IApplicationCommandOption. Rename DefaultPermission to IsDefaultPermission in IApplicationCommand

quinchs committed:
Remove extra white spaces

quinchs committed:
Renamed Joined, Archived, and Locked to HasJoined, IsArchived, and IsLocked

quinchs committed:
Rename Live and DiscoverableDisabled to IsDiscoverableDisabled and IsLive in IStageChannel

quinchs committed:
Remove newline

quinchs committed:
Add indent to summaries

quinchs committed:
Remove unnecessary json serializer field

quinchs committed:
Fix ToEntity for roletags incorrectly using IsPremiumSubscriber

quinchs committed:
Update RestChannel for new channel types

quinchs committed:
Fix different rest channels not deserializing properly

quinchs committed:
fully qualify internal for UrlValidation and add indent to summary

quinchs committed:
Add missing periods to InteractionResponseType

quinchs committed:
Fix summary in IApplicationCommandOptionChoice

quinchs committed:
Update IApplicationCommandOption summaries

quinchs committed:
Update IApplicationCommandInteractionDataOption summaries

quinchs committed:
Update IApplicationCommandInteractionData summaries

quinchs committed:
Update IApplicationCommand summaries

quinchs committed:
Update ApplicationCommandType summaries

quinchs committed:
rename DefaultPermission to IsDefaultPermission in ApplicationCommandProperties

quinchs committed:
update ApplicationCommandOptionChoiceProperties summaries

quinchs committed:
Rename Default, Required, and Autocomplete to IsDefault, IsRequired, and IsAutocomplete in ApplicationCommandOptionProperties

quinchs committed:
Update SlashCommandProperties summaries

quinchs committed:
update SlashCommandBuilder boolean field names, summaries, and choice parameters

quinchs committed:
Update SelectMenuOption summaries, Rename Default to IsDefault in SelectMenuOption

quinchs committed:
update SelectMenuComponent summaries. Rename Disabled to IsDisabled in SelectMenuComponent

quinchs committed:
update ComponentBuilder summaries and boolean fields.

quinchs committed:
Update ButtonComponent summaries and boolean fields

quinchs committed:
update ActionRowComponent summaries

quinchs committed:
Update UserCommandBuilder

quinchs committed:
Update MessageCommandBuilder summaries and boolean properties

quinchs committed:
Update IGuild summary

quinchs committed:
Update IGuild summaries

quinchs committed:
Update StagePrivacyLevel summary

quinchs committed:
update IThreadChannel summaries

quinchs committed:
Update IStageChannel summaries

quinchs committed:
Refactor summaries and boolean property names

d4n3436 committed:
General cleanup (#223)

* General cleanup

* Add Async suffix to SendAutocompleteResult

* Fix more formatting

* Fix unused RequestOptions in GetActiveThreadsAsync

* Add message to ArgumentNullException

quinchs committed:
Ephemeral attachments

quinchs committed:
Add missing jsonproperty attribute

quinchs committed:
Add IMessage.Interaction

quinchs committed:
Update attachment checks for embed urls

quinchs committed:
meta: bump version

quinchs committed:
Remove old package configs and update image

quinchs committed:
Update package logos

quinchs committed:
Fix logo reference for azure

quinchs committed:
Deprecate old package definitions in favor for target file

Co-authored-by: Jared L <48422312+lhjt@users.noreply.github.com>

quinchs committed:
Deprecate old package definitions in favor for target file

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
Update package ids

quinchs committed:
Fix url validation

quinchs committed:
meta: bump version

emillly-b committed:
Fix assignment of UserMentions (#233)

emillly-b committed:
Fix CleanContent (#231)

emillly-b committed:
Fix SocketSlashCommandData access modifier. (#237)

Fixes #229

csmir committed:
Update README with better header (#232)

* Update README with better header

Adds HTML elements that implement the main logo & improve the redirection tag positions.

* Resolving border issue in light-mode

* Update sponsor section

quinchs committed:
Implement checks for interaction respond times and multiple interaction responses. closes #236, #235

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
Add response check to socket auto complete

quinchs committed:
meta: bump versions

quinchs committed:
Fix #239

quinchs committed:
meta: bump version

quinchs committed:
meta: update logo

quinchs committed:
meta: bump versions

quinchs committed:
Revert received at time, confirmed by discord staff to be accurate

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into merger-labs

Update requested changes of obsolete and references to labs.

Added `Interaction` to `IMessage`
Fixed grammar
Fixed bugs relating to interactions.

quinchs committed:
Update docs

quinchs committed:
Update CHANGELOG.md

quinchs committed:
meta: docs building

quinchs committed:
Update docs.yml

quinchs committed:
Update docs.yml

quinchs committed:
Fix docfx version

quinchs committed:
Update docs.yml

quinchs committed:
Update docs.bat

quinchs committed:
Rename docs repo for clone

quinchs committed:
update docfx version

quinchs committed:
Update docs.bat

quinchs committed:
Update docfx version

quinchs committed:
Remove docs from pipeline

csmir committed:
FAQ revamped, metadata updated (#241)

* FAQ revamped, metadata updated

* Update FAQ.md

* Update README.md

* Docs index improvement

quinchs committed:
Fix InvalidOperationException in modify channel

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
feature: guild avatars, closes #238

quinchs committed:
feature: modify role icons

quinchs committed:
meta: changelog

quinchs committed:
meta: bump version

quinchs committed:
Update README.md

quinchs committed:
Fix non value type options not being included in autocomplete

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

BillChirico committed:
Add new activity flags (#254)

* Add new activity flags

* Add missing commas

WilliamWelsh committed:
Added support for GUILD_JOIN_REQUEST_DELETE event (#253)

Fixes #247

drobbins329 committed:
Adding BotHTTPInteraction user flag (#252)

MrCakeSlayer committed:
animated guild banner support (#255)

csmir committed:
Docs work (WIP) (#242)

* Main page work

* Metadata logo dir

* More main page edits

* Naming change

* Dnet guide entries pruned

MrCakeSlayer committed:
Add student hub guild directory channel (#256)

* animated guild banner support

* Add guild directory channel

quinchs committed:
Fix followup with file overwrite having incorrect parameter locations

quinchs committed:
Merge labs 3.x

quinchs committed:
Update GUILD_JOIN_REQUEST_DELETE event

quinchs committed:
Update head.tmpl.partial

Liege72 committed:
Removed BannerId and AccentColor  (#260)

* Removed BannerId property, GetBannerURL method, and AccentColor property from IUser and socket entities.

* Fixed errors in IUser.cs

* Added back summary for GetAvatarUrl method in IUser.cs

CottageDwellingCat committed:
Support Guild Boost Progress Bars (#262)

* Support Guild Boost Progress Bars

* Update SocketChannel.cs

* Fix non-optional and unnecessary values.

* Spelling

* Reordering and consistency.

quinchs committed:
Remove log for reconnect

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

CottageDwellingCat committed:
Add missing flags to SystemChannelMessageDeny (#267)

quinchs committed:
Fix labs reference in analyzer project and provider project

quinchs committed:
Rename new activity flags

quinchs committed:
Guild feature revamp and smart gateway intent checks

quinchs committed:
Get thread user implementation

drobbins329 committed:
Amend creating slash command guide (#269)

* Adding BotHTTPInteraction user flag

* Added comments explaining the Global command create stipulations.

quinchs committed:
Fix numeric type check for options

CottageDwellingCat committed:
Add state checking to ConnectionManager.StartAsync (#272)

quinchs committed:
initial interface changes

quinchs committed:
Multi file upload + attachment editing

quinchs committed:
meta: bump versions

quinchs committed:
Update CHANGELOG.md

quinchs committed:
Update CHANGELOG.md

CottageDwellingCat committed:
Support Min and Max values on ApplicationCommandOptions (#273)

* Support Min and Max values on ApplicationCommandOptions

* Support decimal min/max values

* Docs imrpovments + use ToNullable

csmir committed:
Logomark, doc settings edit (#258)

* Logomark, doc settings edit

* Replace standard logo

* Bumping docfx plugins to latest release

* Bump version metadata

* Logo svg fix

quinchs committed:
Change default sticker behavior and add AlwaysResolveSticker to the config

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
Implement rest based interactions. Added ED25519 checks. Updated summaries.

daanvandenhoek commented:
Any ETA on this?

quinchs committed:
Update package logo

CottageDwellingCat committed:
Automatically fix ordering of optional command options (#276)

* auto fix optional command option order

* clean up indentation

Floowey committed:
Fix maximum number of Select Menu Options (#282)

As of https://discord.com/developers/docs/interactions/message-components#select-menu-object-select-menu-structure the maximum number of options is 25, not less than 25. Hopefully the change catches all necessary locations

quinchs committed:
Add voice region to modify voice channels

quinchs committed:
Update summaries on rest interactions

quinchs commented:
*Hopefully* sometime within the next 7 days if everything goes according to plan

Cenngo committed:
Interaction Specific Interfaces (#283)

* added interaction specific interfaces

* fix build error

* implement change requests

quinchs committed:
Update application

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
Add Guild Scheduled Events (#279)

* guild events initial

* sharded events

* Add new gateway intents and fix bugs

* More work on new changes to guild events

* Update guild scheduled events

* Added events to extended guild and add event start event

* Update preconditions

* Implement breaking changes guild guild events. Add guild event permissions

* Update tests and change privacy level requirements

* Update summaries and add docs for guild events

quinchs committed:
meta: bump version

Lojemiru commented:
Looking forward to it! I have a single-server administration bot written in discord.py and have to migrate while keeping an eye out for slash commands, and .NET's my preferred working environment. It's very reassuring to see that these and a plethora of other improvements/features are coming in a relatively short timespan.

csmir committed:
Increment meta version (#285)

* Increment meta version

* Update docfx.json

quinchs committed:
Fix #289 and add configureawaits to rest based interactions

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
meta: bump version

quinchs committed:
Add GUILD_SCHEDULED_EVENT_USER_ADD and GUILD_SCHEDULED_EVENT_USER_REMOVE (#287)

quinchs committed:
Remove newline

quinchs committed:
Fix autocomplete result value

quinchs committed:
meta: bump versions

quinchs committed:
Add `GuildScheduledEventUserAdd` and `GuildScheduledEventUserRemove` to sharded client

d4n3436 committed:
Make RestUserCommand public (#292)

exsersewo committed:
Fix Components not showing on FUWF (#288) (#293)

Adds Components to Payload JSON Generation

Pythonic-Rainbow commented:
> Looking forward to it! I have a single-server administration bot written in discord.py and have to migrate while keeping an eye out for slash commands, and .NET's my preferred working environment. It's very reassuring to see that these and a plethora of other improvements/features are coming in a relatively short timespan. Thank you all so much for this huge effort.

Yeah I am also planning to port my bot to Discord.NET  from .py as well haha, this is going to be a great chance to improve my C# too

quinchs committed:
Implement smarter rest resolvable interaction data. Fixes #294

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
Add UseInteractionSnowflakeDate to config #286

quinchs committed:
Implement Better Discord Errors (#291)

* Initial error parsing

* Implement better errors

* Add missing error codes

* Add voice disconnect opcodes

* Remove unused class, add summaries to discordjsonerror, and remove public constructor of slash command properties

* Add error code summary

* Update error message summary

* Update src/Discord.Net.Core/DiscordJsonError.cs

Co-authored-by: Jared L <48422312+lhjt@users.noreply.github.com>

* Update src/Discord.Net.WebSocket/API/Voice/VoiceCloseCode.cs

Co-authored-by: Jared L <48422312+lhjt@users.noreply.github.com>

* Fix autocomplete result value

Co-authored-by: Jared L <48422312+lhjt@users.noreply.github.com>

pdelvo committed:
Change the minimum length of slash commands to 1 (#284)

* Change the minimum length of slash commands to 1. This is the correct value according to the docs and it has been changed after user feedback.

* Fix the limit in 3 other places

Co-authored-by: quin lynch <lynchquin@gmail.com>

quinchs committed:
Add new thread creation properties

quinchs committed:
Add role emoji. Fixes #295

quinchs committed:
Fix mocked text channel

quinchs committed:
Fix precondition checks. Closes #281

quinchs committed:
Initial fix (#297)

quinchs committed:
meta: bump version

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into release/3.x

quinchs committed:
Merge branch 'release/3.x' of https://github.com/Discord-Net-Labs/Discord.Net-Labs into merger-labs

quinchs committed:
Update from release/3.x

quinchs commented:
## Updates
- Added Emoji to roles
- Added Invitable and Slowmode to thread creation
- Added better discord errors.
- Added UseInteractionSnowflakeDate to config
- Make RestUserCommand public https://github.com/Discord-Net-Labs/Discord.Net-Labs/pull/292
- Fix Components not showing on FollowUpWithFile https://github.com/Discord-Net-Labs/Discord.Net-Labs/pull/293
- Implement Better Discord Errors https://github.com/Discord-Net-Labs/Discord.Net-Labs/pull/291
- Change the minimum length of slash commands to 1 https://github.com/Discord-Net-Labs/Discord.Net-Labs/pull/284
- Fix ratelimit timings https://github.com/Discord-Net-Labs/Discord.Net-Labs/pull/297
- Fixed preconditions using less than when they should be at most
- Fixed rest resolvable data being lazy
- Changed the minimum length of slash command names to 1
- Fixed autocomplete not accepting values
- Added GuildScheduledEventUserAdd and GuildScheduledEventUserRemove
- Added new Cacheable<> type for rest based downloads while maintaining strong typed entities.
- Slash commands will now have populated data.
- Added ConfigureAwait(false) to async calls in rest based interactions
- Added guild events.
- Added Rest-based http interactions.
- Added RTCRegion to voice channel properties.
- Changed default sticker behavior. You can configure getting default stickers on startup and auto resolve stickers with AlwaysDownloadDefaultStickers and AlwaysResolveStickers. 
- Support Min and Max values on ApplicationCommandOptions https://github.com/Discord-Net-Labs/Discord.Net-Labs/pull/273
- Logomark, doc settings edit https://github.com/Discord-Net-Labs/Discord.Net-Labs/pull/258
- Automatically fix ordering of optional command options https://github.com/Discord-Net-Labs/Discord.Net-Labs/pull/276
- Fix maximum number of Select Menu Options https://github.com/Discord-Net-Labs/Discord.Net-Labs/pull/282
- Interaction Specific Interfaces https://github.com/Discord-Net-Labs/Discord.Net-Labs/pull/283
- Add Guild Scheduled Events https://github.com/Discord-Net-Labs/Discord.Net-Labs/pull/279
- Added GetThreadMemberAsync implementation.
- Added GuildFeatures to IGuild.
- Added checks for gateway intent in some methods. DownloadUsersAsync will throw an exception if you don't have the gateway intent enabled locally now, this should help new people specify the correct intents locally.
- Added strong types to GuildJoinRequestDeleted events.
- Added SendFilesAsync to channels.
- Added Attachments property to MessageProperties
- Fixed numeric type check for slash command options.
- Fixed followup with file overwrite having incorrect parameter locations.
- Fixed non value type options not being included in autocomplete.
- Renamed PARTY_PRIVACY_FRIENDS, PARTY_PRIVACY_VOICE_CHANNEL, and EMBEDDED to PartyPrivacyFriends, PartyPrivacyVoiceChannel, and Embedded in ActivityProperties.
- Removed error log for gateway reconnects.

quinchs committed:
Remove more labs references

quinchs committed:
Remove doc file for Discord.Net.Analyzers

kieranedwards commented:
Hi, it appears that this PR has been merged into the "Dev" branch but I do not see new packages being generated in NuGet package manager.

Does anyone know what we need to do to get the updated packages published?

Many thanks!

quinchs commented:
> Hi, it appears that this PR has been merged into the "Dev" branch but I do not see new packages being generated in NuGet package manager.
> 
> Does anyone know what we need to do to get the updated packages published?
> 
> Many thanks!

Theres still alot to do before we can release a nuget build, you can look at the projects tab and see what needs to be done, I also need to merge in a slash command service so maybe in a week we will have a nuget build.

kieranedwards commented:
> > Hi, it appears that this PR has been merged into the "Dev" branch but I do not see new packages being generated in NuGet package manager.
> > Does anyone know what we need to do to get the updated packages published?
> > Many thanks!
> 
> Theres still alot to do before we can release a nuget build, you can look at the projects tab and see what needs to be done, I also need to merge in a slash command service so maybe in a week we will have a nuget build.

Ah! Of course, my apologies, I hadn't realised. Thank you very much! :)

RheaAyase commented:
This should have been rebased to be able to find related changes between files per commit... This is really horrible merge commit T_T
